### ISSUE TITLE
decode() on bytes should support UTF-16

### ISSUE THREAD

**Issue by sethhall** [zeek/spicy#1788]
It looks like the current implementation only supports ASCII and UTF-8 to decode into a string and the current library being used is strictly for UTF-8.  In order to support anything with Windows roots, it would be nice to support UTF-16.

I poked around for a few minutes and found a potential small library that might work for the use case to decode UTF-16 into a string type.... https://github.com/nemtrif/utfcpp

**Comment by sethhall** [comment]
This still needs to be done (because the decode() method was clearly built with this in mind, but as a stop gap, I have a UTF-16 string reader (and it converts to utf-8 internally) implemented natively in spicy here: https://github.com/sethhall/spicy-parsers/blob/main/unicode/utf16.spicy

**Comment by Ethanholtking** [comment]
Hello I'd like to help solve this issue, however, I'm having difficulty trying to find where exactly the problem is. Could someone please help me locate where the decode() is?

**Comment by bbannier** [comment]
> Hello I'd like to help solve this issue, however, I'm having difficulty trying to find where exactly the problem is. Could someone please help me locate where the decode() is?

Implementing the runtime part would go roughly like the following:

1. Adding a `UTF16` `Charset` value here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/hilti/runtime/include/types/bytes.h#L42
2. Implementing handling of `Charset::UTF16` in `Bytes::decode` here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/hilti/runtime/src/types/bytes.cc#L105-L132 The C++ unit test for `Bytes::decode` should also be updated here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/hilti/runtime/src/tests/bytes.cc#L64-L83

To make this available in Spicy code it needs to be added to both HILTI as well as Spicy:

1. Add it to HILTI here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/hilti/lib/hilti.hlt#L14 There should also be a new test case here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/tests/hilti/types/bytes/decode.hlt
2. Add it to Spicy here: https://github.com/zeek/spicy/blob/943dea8d284c3b6fd65426e6e22abce1669ceeb1/spicy/lib/spicy.spicy#L33-L37 Adding a test case is not strictly needed since this just wraps HILTI functionality.

**Comment by rsmmr** [comment]
@Ethanholtking Did that help? Are you working on this?

**Comment by bbannier** [comment]
I looks like even with https://github.com/nemtrif/utfcpp we might not be able to ditch our dependency on https://github.com/nemtrif/utfcpp. `utfcpp` provides UTF-8, UTF-16, and UTF-32 and functions to work on them and convert between the types, but for functions like `upper`/`lower` we still need `utf8proc`.

I took a stab at an implementation, and this looks feasible. Our current model is that `bytes` are just bags of bytes while `string` is valid UTF-8. It might be that the semantics of `string`-like functions in `bytes` become muddy by introducing UTF-16 (functions like `startsWith`/`upper`/`lower`). Since recently with #1828 we introduce similar functions for `string` it might be that they can and should completely go away in `bytes`.

**Comment by bbannier** [comment]
> I think your issues begs the question of if it makes sense to support anything other than UTF-8? What I mean by this is that maybe UTF-16/32 support should be limited to the decode function on bytes? We could ensure that the resulting string is still only a UTF-8 string internally and externally to avoid the potential conflicts you're mentioning.
> 
> When I originally filed this ticket, that's what I had in mind at least.

That was also what I had in mind. We'll keep `bytes` to be a bag of bytes, and `string` always an UTF-8 string. We will allow interpreting of UTF-8 `bytes` via `decode`, but the result will stay an UTF-8 string.[^1]

[^1]: Otherwise a string would need to know its own encoding which I suspect could make the majority of existing use cases more expensive. It might also need to be able to store an UTF-16 string internally; right now `string` is just a `std::string` in the backend, and optionally allowing it to be backed by a `std::wstring` could introduce more costs.

The issue around the `bytes` helper function I mentioned is related, e.g., `startsWith` could need to interpret the bytes as UTF-16, or one would need to remember that a `bytes` produced by e.g., `bytes::upper` might be UTF-16. This seems to make complicated for no good reason. It might be easier to just force using e.g., `string::upper` instead which is always UTF-8; similarly `startsWith` would have a clear meaning for a `string` while for `bytes` it is more like a check on individual `bytes`. Not a very pressing issue, but probably a worthwhile cleanup.

**Comment by rsmmr** [comment]
> string-like functions in bytes become muddy by introducing UTF-16 (functions like startsWith/upper/lower)

Don't they'd get muddy. `startsWith()` takes a `bytes` argument, so that's really just a byte-for-byte comparison without any character set interpretation at all. `upper()/lower() receive the encoding as an argument, which should just generalize to UTF-16, no? And in fact their implementation is really just a shortcut for going through a string manually.

**Comment by bbannier** [comment]
> > string-like functions in bytes become muddy by introducing UTF-16 (functions like startsWith/upper/lower)
> 
> I don't think they'd get muddy. `startsWith()` takes a `bytes` argument, so that's really just a byte-for-byte comparison without any character set interpretation at all. `upper()/lower()` receive the encoding as an argument, which should just generalize to UTF-16, no? And in fact their implementation is really just a shortcut for going through a string manually.

Yes, this it is definitely doable. What I meant is: assume you start out with UTF-16 `bytes`; calling to `upper` with `cs=UTF16` on this would return a new `bytes` which again should be `UTF-16`. The implementation create a temporary UTF-8 `string`, invoke `string::upper` on this, convert this back to `UTF-16`. The intermediary conversion to UTF-8 is since case conversions would still need to be handled with utf8proc which does only supports UTF-8.

**Comment by rsmmr** [comment]
> The intermediary conversion to UTF-8 is since case conversions would still need to be handled with utf8proc which does only supports UTF-8.

Ah, ok, so it's "muddy" in the implementation sense, not in the API sense. That seems fine.
